### DISCUSSION TITLE
Airflow task fails when no permission to bash

### DISCUSSION THREAD

**Discussion by ghost** [apache/airflow#38286] [General]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### If "Other Airflow 2 version" selected, which one?

2.5.3

### What happened?

Hello Team,
We are trying to deploy Airflow as non-root user in a secure kubernetes environment. Once the helm chart is deployed the web, scheduler and sync users deployment start failing with error - PermissionError: [Errno 13] Permission denied: 'bash'  We are not allowed to use bash instead we use /bin/sh. So here is what I have understood from stacktrace that whenever a airflow command Ex. airflow version or airflow scheduler is executed it triggers a python script in backend which then uses subprocess.popen to run some shell command and its using bash as default. Is there a way we can override it to use /bin/sh ?
Any help on this is highly appreciated. Thanks in advance

### What you think should happen instead?

To run with the allowed shell /bin/sh

### How to reproduce

remove the permission for the /bin/bash in kubernetes

### Operating System

Debian GNU 11 (bullseye)

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon==7.3.0
apache-airflow-providers-apache-spark==1.0.3
apache-airflow-providers-celery==3.1.0
apache-airflow-providers-cncf-kubernetes==5.2.2
apache-airflow-providers-common-sql==1.3.4
apache-airflow-providers-docker==3.5.1
apache-airflow-providers-elasticsearch==4.4.0
apache-airflow-providers-ftp==3.3.1
apache-airflow-providers-google==8.11.0
apache-airflow-providers-grpc==3.1.0
apache-airflow-providers-hashicorp==3.3.0
apache-airflow-providers-http==4.2.0
apache-airflow-providers-imap==3.1.1
apache-airflow-providers-microsoft-azure==5.2.1
apache-airflow-providers-mysql==4.0.2
apache-airflow-providers-odbc==3.2.1
apache-airflow-providers-postgres==5.4.0
apache-airflow-providers-redis==3.1.0
apache-airflow-providers-sendgrid==3.1.0
apache-airflow-providers-sftp==4.2.4
apache-airflow-providers-slack==7.2.0
apache-airflow-providers-snowflake==4.0.4
apache-airflow-providers-sqlite==3.3.1
apache-airflow-providers-ssh==3.5.0

### Deployment

Other 3rd-party Helm chart

### Deployment details

Kubernetes version 1.16

### Anything else?

Everytime we deploy in a secure environment.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by hussein-awala** [comment]
Are you using the official docker image? if yes, could you provide the hash of the image?

**Comment by ghost** [comment]
@hussein-awala We are using apache/airflow:2.5.3-python3.8. The hash is not matching as we are adding some more packages to the image and copying some dags as well.

**Comment by potiuk** [comment]
The scripts you mentioned are using bash and this is what is going to stay until somone (you?) rewrites them. If you have some internal way to prevent it, you can always extend the image and replace the scripts to use `/bin/sh` instead, so you are not blocked. 

But this will likely be a lot of changes as we use bash as a shell of choice internally for everything, including the fact that if you extend the image by default all the docker command are using bash for docker RUN command.

It's quite a huge undertaking to rewrte everything to /bin/sh - we already decreased our shell reliance a lot and most of our scripts  are now Python, but we stuill havce 2307 lines of bash in our scripts directory (we used to have > 13000).

You might want to take on the task of rewrting those in Python - happy to accep PRs. Rewriting the scripts to `/bin/sh` instead would be rather useless as we want to get rid of the shell code altogether. Occasionally people contribute those and you and your team can also do so - see for example https://github.com/apache/airflow/pull/38278  - you can similarly rewrite those scripts we have (or wait until the volunteers from all over the place will do it).

Happy to accept PRs to do so. You can then contribute back for the free software you get. 

Or (alternatively) you should allow for bash execution. That's a strange requirement you have there to block bash from execution, never heard of it - so you are likely alone with that problem and you will need to extend the images yourself until we have all our bash reliance removed. Probably it will be a lot easier for you to ask for exception than to do the rewrites yourself (and contribute it back) - those are the two choices you basically have.

Here are the current numbers for lines of code:

``
SLOC	Directory	SLOC-by-Language (Sorted)
297102  tests           python=296930,xml=132,sh=40
208574  airflow         python=205893,javascript=2681
60666   build           python=57947,javascript=2719
27514   dev             python=27358,sh=156
27451   clients         python=27416,sh=35
13577   docs            javascript=9522,python=3766,sh=289
12501   helm_tests      python=12501
9691    scripts         python=7304,sh=2387
1549    kubernetes_tests python=1549
455     docker_tests    python=455
113     top_dir         python=113
64      chart           sh=64
```

**Comment by potiuk** [comment]
Converting to dicussion in case more discussion is needed.
